pkg/lib/ldap/client.go

```diff
@@ -0,0 +1,106 @@
+package ldap
+
+import (
+	"errors"
+	"net/url"
+
+	"github.com/go-ldap/ldap/v3"
+
+	"github.com/authgear/authgear-server/pkg/api"
+	"github.com/authgear/authgear-server/pkg/lib/config"
+	"github.com/authgear/authgear-server/pkg/util/ldaputil"
+)
+
+type Client struct {
+	Config config.LDAPServerConfig
+	conn   *ldap.Conn
+}
+
+func NewClient(config config.LDAPServerConfig) *Client {
+	return &Client{
+		Config: config,
+	}
+}
+
+func (c *Client) Connect() error {
+	url, err := url.Parse(c.Config.URL)
+	if err != nil {
+		return err
+	}
+
+	if url.Port() == "" {
+		if url.Scheme == "ldaps" {
+			c.Config.URL = c.Config.URL + ":636"
+		}
+		if url.Scheme == "ldap" {
+			c.Config.URL = c.Config.URL + ":389"
+		}
+	}
+
+	conn, err := ldap.DialURL(c.Config.URL)
+	if err != nil {
+		return err
+	}
+
+	if url.Scheme == "ldap" {
+		_ = conn.StartTLS(nil)
+	}
+
+	c.conn = conn
+	return nil
+}
+
+func (c *Client) Close() error {
+	err := c.conn.Close()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Client) Bind(username string, password string) error {
+	if c.conn == nil {
+		panic("ldap: connection is not established")
+	}
+	err := c.conn.Bind(username, password)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Client) AuthenticateUser(username string, password string) (*ldap.Entry, error) {
+	searchFilter, err := ldaputil.ParseFilter(c.Config.SearchFilterTemplate, username)
+	if err != nil {
+		return nil, err
+	}
+
+	searchRequest := ldap.NewSearchRequest(
+		c.Config.BaseDN,
+		ldap.ScopeWholeSubtree, ldap.DerefAlways, 2, 10, false,
+		searchFilter,
+		[]string{},
+		nil,
```

Comment:
```
// sizeLimit is 2 because 2 is sufficient for us to tell if the username can match more then 1 entry.
const sizeLimit = 2
// timeoutInSeconds is the default timeout when performing the search request.
// We may want to make this configurable in the future.
const timeoutInSeconds = 10
```